src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/AppModel/ResourceContainer.cs

```diff
@@ -243,123 +184,111 @@ protected override PackagePart GetPartCore(Uri uri)
-        // <returns></returns>
-        private ResourceManagerWrapper GetResourceManagerWrapper(Uri uri, out string partName, out bool isContentFile)
+        /// <summary>
+        /// Searches the available ResourceManagerWrapper list for one that matches the given Uri.
+        /// It could be either ResourceManagerWrapper for specific library assembly or Application
+        /// main assembly. Package enforces that all Uri will be correctly formatted.
+        /// </summary>
+        /// <param name="uri">Assumed to be relative</param>
+        /// <param name="partName">The name of the file in the resource manager</param>
+        /// <param name="isContentFile">A flag to indicate that this path is a known loose file at compile time</param>
+        /// <returns></returns>
+        private static ResourceManagerWrapper GetResourceManagerWrapper(Uri uri, out string partName, out bool isContentFile)
         {
-            string assemblyName;
-            string assemblyVersion;
-            string assemblyKey;
             ResourceManagerWrapper rmwResult = ApplicationResourceManagerWrapper;
-
             isContentFile = false;
 
-            BaseUriHelper.GetAssemblyNameAndPart(uri, out partName, out assemblyName, out assemblyVersion, out assemblyKey);
+            BaseUriHelper.GetAssemblyNameAndPart(uri, out partName, out string assemblyName, out string assemblyVersion, out string assemblyToken);
 
-            if (!String.IsNullOrEmpty(assemblyName))
+            if (!string.IsNullOrEmpty(assemblyName))
             {
-                string key = assemblyName + assemblyVersion + assemblyKey;
-
-                _registeredResourceManagers.TryGetValue(key.ToLowerInvariant(), out rmwResult);
-
-                // first time. Add this to the hash table
-                if (rmwResult == null)
+                // Create the key, in format of $"{assemblyName}{assemblyVersion}{assemblyToken}"
+                int totalLength = assemblyName.Length + assemblyVersion.Length + assemblyToken.Length;
+                Span<char> key = totalLength <= 256 ? stackalloc char[totalLength] : new char[totalLength];
+                assemblyName.AsSpan().ToLowerInvariant(key);
+                assemblyVersion.CopyTo(key.Slice(assemblyName.Length));
+                assemblyToken.CopyTo(key.Slice(assemblyName.Length + assemblyVersion.Length));
+
+                // First time; add this to the dictionary and create the wrapper if its not the application entry assembly
+                if (!s_registeredResourceManagersLookup.TryGetValue(key.Slice(0, totalLength), out rmwResult))
                 {
-                    Assembly assembly;
-
-                    assembly = BaseUriHelper.GetLoadedAssembly(assemblyName, assemblyVersion, assemblyKey);
-
+                    Assembly assembly = BaseUriHelper.GetLoadedAssembly(assemblyName, assemblyVersion, assemblyToken);
                     if (assembly.Equals(Application.ResourceAssembly))
```

Comment:
Seems this is run only once and then stored in the cache of resource managers.

Reply:
Speedscope, is that what the cool kids use now? Haha.

Anyways as @miloush says, this is usually only run once, you shall not make it past the lookup, I originally had string there for the key but changed it in the last commits as you rarely need to add an entry.

The method itself shall be optimized but I felt it doesn't need to be done in the scope of this PR, however any modification has to wait anyways. I currently have 60 open PRs and I'm waiting for them to be resolved so I can make more changes now because I'm effectively blocking myself in a lot of places to not overlap changes.

A bit frustrating but gotta live with the reality.

Reply:
> I currently have 60 open PRs and I'm waiting for them to be resolved so I can make more changes now because I'm effectively blocking myself in a lot of places to not overlap changes.
> 
> A bit frustrating but gotta live with the reality.

Please post the PRs which are blocking (causing overlap), We will try to prioritize those. The next CTP has 5 of your PRs lined up. Agree with the sentiment that this is a slow pace and frustrating, however the test passes take up time, and with the .NET 9 releases and other deliverables, the max we can get is 8 PRs a month.

Reply:
@singhashish-wpf would it be possible to run the tests on a batch of PRs together? Or perhaps @h3xds1nz could do a bigger PR with all the picked commits?

Reply:
We are going to batch in 10. The problem is we want the haystack to be smaller in case of test failures, otherwise we could have taken all of them together.

Reply:
@singhashish-wpf Thank you, that would be very lovely if we could prioritize. I understand the amount of tests run is rather large for WPF, it takes time (was it even a few days?) and you're trying to minimize the number so it is easier to pinpoint in case of encountering case failures. Nonetheless, I'm very glad you're taking the time for the discussion here.

As per the priorities, this would help a bunch:

1) #9363 
- This will unblock me on rather a big PR in terms of allocations with most converters (but small in code changes).
2) #9676 / #9697 and #9683
-  This will unblock optimizations of Key/MouseGesture converters as a result.
3) #9220
- I keep constantly hitting on code with changes I made here.
4) #9711 
- Here I'd mostly like to know ahead if we're gonna go that direction, because then I could add the dead code removal that will end up as a result directly into the PR. The save of that is like 12ms during startup.
5) #9481
- This would be nice to have faster for further optimizations on the call-chains I had in mind.
6) #9292
- This PR is from Thomas but would help a great deal to be squeezed in.
7) #9521
- This directly makes a modification in the `BaseUriHelper` for example. The goal is to get rid of the CAS helpers as it should have been done few years ago before your team has stepped into WPF.

I'd also have one question - since I've been mostly writing tests for all the changes, should I be adding them in? My issue is that currently it is hard to not duplicate some tests from wpf-test (any movement on the intended move of the wpf-test repo into main repo as per the discussion in #9401?) and few assemblies don't even have Test projects, the assemblies would require `InternalsVisibleTo` or at least use of UnsafeAccessors/reflection ain't the best idea, I know some DRTs still depend on it though.

Reply:
Thanks @h3xds1nz for the list. I will get these added in the next CTP. cc:// @harshit7962 

If you are writing unit tests, we still have some in the System.Xaml area and others, you can add them in this repo only. If you want to leverage the DRT setup, you can do it on wpf-test repo. We are looking for better ways to have a common place and an integrated testing infra. Tests at both the places are validated before PR merge, so as of now submitting at any place is fine. The gist of the discussion is hinting towards movement of the tests from wpf-test repo, which is going to take up time/effort, also we ideally don't want to keep following the old testing infra as that has execution challenges and learning curve for contributors.

Reply:
@singhashish-wpf Thanks a bunch. Yeah not keen on the DRT setup, thanks for the guidance regarding unit tests.

Happy to hear that things are moving in some direction :)